Ember 4.0 release blog post

jenweber opened the pull request:
This is the starting place for the blog post. Content needs to be filled in by other team members.

**Sign-off:**
- [x] ember-data (@mixonic)
- [x] ember-cli (@mixonic)
- [x] ember.js (@mixonic)
- [x] learning team (@mixonic)

**When to merge**
- After all teams have reviewed/signed off
- After Guides & API docs for v4 are live. Please @ the learning team.

Closes https://github.com/ember-learn/ember-blog/issues/1013

jenweber committed:
Begin drafting the 4.0 release blog post

jenweber commented:
@mixonic do you have anything in progress that you can push up? Feel free to add to this branch directly.

jenweber committed:
Rough draft of 4.0 post

jenweber commented:
@kategengler @rwjblue I just pushed up a very rough draft. Please review for accuracy and general organization. I am especially looking for feedback about whether to include the major version soapbox. I can edit to to make it shorter. I'm thinking of writing a separate post about release strategy, and would move the bulk of it there. Thank you!

jenweber commented:
My to-dos:
- fix typos
- Make spellchecker happy
- Add a section with where to get help
- Link to Octane

jenweber committed:
Add a help section

NullVoxPopuli commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,231 @@
+Ember CLI. Ember it focuses
+the framework by removing long-deprecated APIs and support for legacy platforms.
+With these changes, Ember Octane becomes the norm!
+As planned, this release introduces no new features.
```
This is orthogonal to the release post, but I feel like we need a subclass of semver to communicate our sort of take on this. Some projects just do breaking changes whenever they feel like and never deprecate things, and they burn through majors like it's their job. :thinking: Maybe @wycats' wordsmithery has some ideas <3

NullVoxPopuli replied:
(though, I now see this is expanded on in the next section, but good words (ie: a name) give good marketing, so :shrug: )

jenweber replied:
That's a great point. I think we need a full blog post on this topic. We just need a name for it. I don't think there is one.

mixonic replied:
We should be careful to keep the release post focused on the release. It is dense enough as is, and imo not the place to argue or advocate for release/versioning strategy in depth.

MelSumner replied:
I agree with @mixonic - @NullVoxPopuli if you want to DM me in Discord, I'll help you coordinate this. I'm going to resolve this conversation for now.

NullVoxPopuli commented on content/ember-4-0-released.md:
````diff
@@ -0,0 +1,231 @@
+```bash
+npx ember-cli new my-project
+cd my-project
+npx ember serve # Then visit http://localhost:4200
````
does `npx ember` point at ember-cli?

jenweber replied:
Not exactly? This is invalid:

```
npx ember new my-test-app
npm ERR! could not determine executable to run
```

This _is_ valid.
```
npx ember-cli serve
```

Any thoughts on whether we should be using `npx ember-cli` for all command examples?

mixonic replied:
IMO npx makes sense for installation, but traditional use of the 'ember' command should be the norm after project setup.

mixonic committed:
Additions and adjustments

jenweber commented on content/ember-4-0-released.md:
@mixonic One thing we chatted about in the framework meeting is that somewhere, we should acknowledge the timeline shifts. Any idea of where that should go?

mixonic replied:
We did in an update to the road to 4.0 blog post. I'm not very interested in saying "thanks for your patience", I'd rather direct a closing paragraph with thanks at the many volunteers who worked to make this happen at all.

Maybe it could be mentioned up front, just as a "originally 4.0 was expected on X date"? We can discuss more on Discord as well. I don't want this to be an apology or something though. Lets just be proud to celebrate the excellent work we've done.

jenweber commented on content/ember-4-0-released.md:
```diff
@@ -2,39 +2,48 @@
+Ember 3.15, the framework has continued to support "Classic" framework features
+in accordance with our semantic versioning commitment. Ember 4.0 takes a step
+forward and drops already-deprecated classic APIs, however the foundational `EmberComponent`
+and `EmberObject`/`computed` APIs are not removed in this release.
```
@mixonic this is perfect, thank you!

mixonic committed:
Add 4.x tag

mixonic commented:
FYI I have force pushed this branch after rebase.

kategengler commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,231 @@
+
+This blog post will help you understand how to upgrade and navigate any breaking changes.
+
+Additionally, today the 4.1 beta cycle begins for these same projects. Ember's
```
I'd drop this whole paragraph. We don't always have all the betas out by the time of the release post and we usually consider the cycle starting when Ember.js itself branches. The info also seems superfluous to the post.

kategengler commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,231 @@
+   [ember-cli-deprecation-workflow](https://github.com/mixonic/ember-cli-deprecation-workflow)
+   addon to isolate individual deprecations.
+3. Make sure your app builds successfully.
+4. Install `ember-auto-import` if your app does not already have it:
```
I'm not sure this is sufficient to install `ember-auto-import` v2, as it requires webpack also be installed. The README suggests `npm install --save-dev ember-auto-import webpack` Maybe we should point to the `ember-auto-import` README?

kategengler commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,231 @@
+Ember.js is the core of the Ember framework. It provides routing, rendering, and
+dependency injection features.
+
+Ember.js 4.0 introduces no new public API or deprecations. Instead, it is
```
This isn't true -- there's at least one new deprecation for `assign`

mixonic committed:
Changes to reflect feedback

mixonic committed:
Drop npx from serve command

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+already-deprecated classic APIs, however the foundational `EmberComponent` and
+`EmberObject`/`computed` APIs are not removed in this release.
+
+This blog post will help you understand how to upgrade and navigate any breaking
```
I think this sentence would go better at the end of the opening paragraph `Today the Ember project ... support for legacy platforms."` Currently, the next paragraph, which begins with `Additionally`, doesn't sound right due to this sentence.

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+for major releases to be splashy and full of new features, we think it is more
+important to reduce risks for our developers and the users that depend on them.
+
+Our hope is that that library maintainers across the JavaScript ecosystem
```
```suggestion
Our hope is that library maintainers across the JavaScript ecosystem
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+   Many developers can do this by running
+   `npx ember-cli-update --to 3.28`. More details about how to upgrade your Ember
+   app/addon are in the
+   [general upgrade guide](https://cli.emberjs.com/release/basic-use/upgrading/#upgradinganemberappitself).
```
I think we could reverse the order to make the sentence sound more active. (similarly for lines 104-106)

```suggestion
   `npx ember-cli-update --to 3.28`. The [general upgrade guide](https://cli.emberjs.com/release/basic-use/upgrading/#upgradinganemberappitself) has more details about how to upgrade your Ember app/addon.
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+   Applications that need to upgrade through several versions may want to
+   consider the
+   [ember-cli-deprecation-workflow](https://github.com/mixonic/ember-cli-deprecation-workflow)
+   addon to isolate individual deprecations.
```
```suggestion
   consider isolating individual deprecations by using
   [ember-cli-deprecation-workflow](https://github.com/mixonic/ember-cli-deprecation-workflow).
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+   [ember-cli-deprecation-workflow](https://github.com/mixonic/ember-cli-deprecation-workflow)
+   addon to isolate individual deprecations.
+3. Make sure your app builds successfully.
+4. Install `ember-auto-import` if your app does not already have it follow the
```
```suggestion
4. Install `ember-auto-import` if your app does not already have it. Follow the
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+### `ember-auto-import` is now required
+
+Using Ember 4 or higher requires that the project has version `2.0.0` or later
+of `ember-auto-import` installed. Follow the "How to upgrade" steps above to
```
If possible, I would try adding a hyperlink to the section above.

```md
Follow the steps in [How to upgrade](#some-id...).
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+Why is this change required? `ember-auto-import` supports something we call the
+[v2 addon
+format](https://emberjs.github.io/rfcs/0507-embroider-v2-package-format.html).
+This specification describes how to publish an Ember CLI addon which has minimal
```
```suggestion
This specification describes how to publish an Ember CLI addon that has minimal
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+### APIs Removed in 4.0
+
+Below we've listed some of the most significant API removals in Ember.js 4.0.
+and for an full list of removals, see the [Ember.js 3.x deprecation
```
```suggestion
For a complete list of removals, see the [Ember.js 3.x deprecation
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+
+### Deprecations added in 4.0
+
+Public API deprecations added in Ember 4.0 target removal in Ember 5.0. One
```
I found the first sentence hard to parse and understand. If I understood the gist, I think the following will be easier to read.

```suggestion
Public API, marked as deprecated in Ember 4.0, will be removed in Ember 5.0. One
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+## Changes in Ember Data 4.0
+
+---
+
```
Not sure why `---` appears here and a couple of more times below.

```suggestion
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+
+Ember Data is the official data persistence library for Ember.js applications.
+
+This release removes APIs that were [deprecated in the 3.x cycle](https://deprecations.emberjs.com/ember-data/v3.x). They include:
```
```suggestion
This release removes APIs that were [deprecated in the 3.x cycle](https://deprecations.emberjs.com/ember-data/v3.x). Examples include:
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+## Changes in Ember CLI 4.0
+
+---
+
```
```suggestion
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+instructions, please review the [Ember CLI 4.0.0 release page](https://github.com/ember-cli/ember-cli/releases/tag/v4.0.0).
+
+---
+
```
```suggestion
```

ijlee2 commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,244 @@
+## Major Releases in Ember
+
+Why are there no new features in this release?
+This has been the tradition for all major Ember releases since v2.
```
A nitpick. The word `all` may sound weird because there were only 2 major versions before (not several). I think it's okay to leave the word out.

```suggestion
This has been the tradition for major Ember releases since v2.
```

mixonic committed:
Bump date, mention ember-cli-htmlbars in steps

mixonic committed:
Feedback from @ijlee2

mixonic commented:
Thank you @ijlee2! As always I really, really appreciate your review. :bow:

jenweber committed:
Learning team edits

Typo fix, repair commands,
make the major versions sections shorter

jenweber commented:
@mixonic I just pushed a commit with a few changes:

- I shortened the section on major releases. We will offer more detailed info in the release page of the website someday.
- We added `npx` back in front of ember serve. The command will not succeed otherwise, unless someone already has ember cli globally installed.
- We added a note about making sure people are on the latest npm before trying to generate a project. This is to avoid the bug described here: https://github.com/ember-cli/ember-cli/issues/9711

jenweber commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,240 @@
+- `sendAction` is removed in favor of calling closure actions like any other callback. [Guide here](https://deprecations.emberjs.com/v3.x/#toc_ember-component-send-action).
+- `willTransition` and `didTransition` are removed in favor of router service events. [Guide here](https://deprecations.emberjs.com/v3.x/#toc_deprecate-router-events).
+- Computed Property `volatile()` calls are removed in favor of native getters. [Guide here](https://deprecations.emberjs.com/v3.x/#toc_computed-property-volatile).
+- `this.$()` and other jQuery APIs are deprecated in favor of native browser equivalents. [Guide here](https://deprecations.emberjs.com/v3.x/#toc_jquery-apis). An optional feature which restored this and other jQuery-specific features is also removed. [Guide here](https://deprecations.emberjs.com/v3.x/#toc_optional-feature-jquery-integration).
```
Is this deprecated, or removed in v4? This bullet point should be updated if it has been removed.

MelSumner replied:
Looking at [this PR](https://github.com/emberjs/rfc-tracking/issues/3) I don't think the work was completed, but maybe @rwjblue or @simonihmig can check the list to see if the work was completed and the list just wasn't updated.

simonihmig replied:
The list wasn't updated, will leave a comment there. The work is done. And yes, they have been *removed*!

mixonic replied:
I've updated the text here appropriately. I expect it was left over from a copy-paste job.

kategengler commented on content/ember-4-0-released.md:
````diff
@@ -61,9 +49,12 @@ If you're interested in trying Ember for the
 
 ```bash
-npx ember-cli new my-project
+# Make sure you are the latest npm version
````
```suggestion
# Make sure you are on the latest npm version, 8.x
```

jenweber committed:
Update content/ember-4-0-released.md

Co-authored-by: Katie Gengler <katie@kmg.io>

jenweber committed:
Lint fix

bertdeblock commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,233 @@
+For more details on the changes in Ember Data 4.0, please review the
+[Ember Data 4.0.0 release page](https://github.com/emberjs/data/releases/tag/v4.0.0).
+
+## Changes in Ember CLI 4.0
```
Maybe (some of) these might be worth including in the Ember CLI section?

- https://github.com/ember-cli/ember-cli/pull/9563
- https://github.com/ember-cli/ember-cli/pull/9579
- https://github.com/ember-cli/ember-cli/pull/9595

MelSumner replied:
hmmm. Major versions are supposed to only be about removing deprecated APIs not adding new things. Were these really added in 4.0?

bertdeblock replied:
Yes => https://github.com/ember-cli/ember-cli/blob/master/CHANGELOG.md#v400.

mixonic replied:
I've updated the post to call out the CI provider change (also GitHub Actions is now the default in 4.0) and the exclude/include change.

I left out the pnpm PR as I'm not sure how much that is a step toward pnpm support vs "pnpm just works well everywhere now". I think the latter would be worth a call-out, but highlighting an incremental step might be a bit confusing.

I asked @Turbo87 for feedback on Discord. He could always update this in a followup.

bertdeblock commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,233 @@
+Want to help make Ember 4 a success? Here are some things you can do!
+
+- Blog about your experience
+- If someone helps you with a question, consider positing both the question
```
```suggestion
- If someone helps you with a question, consider posting both the question
```

bertdeblock replied:
Oh wow, it seems `positing` is a word. Sorry if this suggestion is incorrect! I thought it was a typo 😬

MelSumner replied:
I think it's the incorrect use of the word `positing` anyway, so your suggestion should still be considered.

jenweber committed:
Update ember-4-0-released.md

MelSumner commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,243 @@
+The following public API, marked as deprecated in Ember 4.0, will be removed in
+Ember 5.0:
+
+- Use of `Ember.assign` is deprecated. You should replace any calls to `Ember.assign` with `Object.assign` or use the object spread operator. See the [deprecation guide](https://deprecations.emberjs.com/v4.x#toc_ember-polyfills-deprecate-assign) for examples.
```
```suggestion
- Use of `Ember.assign` is deprecated. You should replace any calls to `Ember.assign` with `Object.assign` or use the [object spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax). See the [deprecation guide](https://deprecations.emberjs.com/v4.x#toc_ember-polyfills-deprecate-assign) for examples.
```

MelSumner commented on content/ember-4-0-released.md:
```suggestion
  so that others can learn from it.
```

MelSumner commented on content/ember-4-0-released.md:
```suggestion
- Help your favorite addons to upgrade to Ember 4!
```

MelSumner commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,243 @@
+
+Want to help make Ember 4 a success? Here are some things you can do!
+
+- Blog about your experience
```
```suggestion
- Blog about your experience.
```

MelSumner reviewed (approved):
I've suggested a few changes but they are non-blockers. It would help to have some clarity on whether or not ember-cli really did add new things in 4.0.

mixonic commented on content/ember-4-0-released.md:
```diff
@@ -0,0 +1,243 @@
+npx ember-cli@latest new my-project
+cd my-project
+npx ember serve
+# Then visit http://localhost:4200
```
Please confirm these exact commands work as stated before merge (or someone comment here that they have done so).

bertdeblock replied:
In this example, `npx ember serve` will work, because `ember` is the name of the executable within `/node_modules/.bin`. It's the first time I've seen it like this in any documentation though. Running e.g. `npx ember new` somewhere else will not work as it must be `npx ember-cli new` instead. After creating a new app, Ember CLI suggests to run `yarn start` or `npm start`. So that could be suggested as well instead of `npx ember serve`? And the docs use `ember server`.

mixonic replied:
Yeah I really don't like suggesting `npm ember s` here when it is only using the local path. `npx`, to me, always implies running some thing from the package `ember`. In this case it is running something from "locally installed ember-cli", a usage which I've never seen.

I don't know why we're introducing a new style of usage here and now. I'd like top revert it, again. @jenweber I think you brought it back after an earlier reversion, or had comments elsewhere here.

mixonic replied:
Prior discussion:

* https://github.com/ember-learn/ember-blog/pull/1036#discussion_r753834945
* https://github.com/ember-learn/ember-blog/pull/1036#issuecomment-990034431

I think we agree not using `npx ember s` is the norm. I'm not clear on why @jenweber said `ember s` wouldn't work without the global installed. I thought it would be found from `node_modules/.bin` as well. If it isn't, then what we really want is `yarn run ember s`, etc.

Actually, I think `npm start` is the right thing to do here. I'm going to update to that.

mixonic committed:
Language tweak for jQuery removal

mixonic committed:
Apply suggestions from code review

Co-authored-by: Melanie Sumner <melaniersumner@gmail.com>
Co-authored-by: Bert De Block <bert.deblock@bagaar.be>

mixonic committed:
Mention CLI changes

mixonic committed:
npm start

mixonic committed:
Update date

mixonic commented:
I believe this is actually ready to roll. It has been rebased. Date is set to 2021-12-20.

mixonic committed:
Nested lists are not a markdown thing

Also improve consistency in version number formatting.

mixonic committed:
Format header better

mixonic committed:
Tweak headers
